diff --git a/sys/kern/kern_descrip.c b/sys/kern/kern_descrip.c
index a88e6690..8298d93c 100644
--- a/sys/kern/kern_descrip.c
+++ b/sys/kern/kern_descrip.c
@@ -853,6 +853,17 @@ kern_fcntl(int fd, int cmd, union fcntl_dat *dat, struct ucred *cred)
 		error = VOP_ADVLOCK(vp, (caddr_t)p->p_leader, F_GETLK,
 				    &dat->fc_flock, F_POSIX);
 		break;
+
+	case F_GETPATH:
+		if (fp->f_type != DTYPE_VNODE) {
+			kprintf("kern_fcntl() not vnode\n");
+			error = EBADF;
+			break;
+		}
+		vp = (struct vnode *)fp->f_data;
+		error = vn_fullpath(p, vp, &dat->fc_path.ptr, &dat->fc_path.buf, 1);
+		break;
+
 	default:
 		error = EINVAL;
 		break;
@@ -920,6 +931,11 @@ sys_fcntl(struct fcntl_args *uap)
 			error = copyout(&dat.fc_flock, (caddr_t)uap->arg,
 			    sizeof(struct flock));
 			break;
+		case F_GETPATH:
+			error = copyout(dat.fc_path.ptr, (caddr_t)uap->arg,
+				MAXPATHLEN);
+			kfree(dat.fc_path.buf, M_TEMP);
+			break;
 		}
 	}
 
diff --git a/sys/sys/fcntl.h b/sys/sys/fcntl.h
index aa789738..df2d5df2 100644
--- a/sys/sys/fcntl.h
+++ b/sys/sys/fcntl.h
@@ -208,6 +208,7 @@
 #if __BSD_VISIBLE
 #define	F_DUP2FD_CLOEXEC 18		/* Like F_DUP2FD with FD_CLOEXEC set */
 #endif
+#define F_GETPATH	19		/* retrieve full path to file associated with fd */
 
 /* file descriptor flags (F_GETFD, F_SETFD) */
 #define	FD_CLOEXEC	1		/* close-on-exec flag */
@@ -243,6 +244,10 @@ union fcntl_dat {
 	int		fc_flags;	/* F_GETFL/F_SETFL */
 	int		fc_owner;	/* F_GETOWN/F_SETOWN */
 	struct flock	fc_flock;	/* F_GETLK/F_SETLK */
+	struct {
+		char* 	ptr;
+		char* 	buf;
+	} fc_path; 			/* F_GETPATH */
 };
 #endif /* _KERNEL */
 
